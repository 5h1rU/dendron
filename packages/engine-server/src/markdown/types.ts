import { DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
  children: MDAST.Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  // MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
};

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};
