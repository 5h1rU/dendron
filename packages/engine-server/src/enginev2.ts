import {
  ConfigWriteOpts,
  DendronConfig,
  DendronError,
  DEngineClientV2,
  DEngineDeleteSchemaRespV2,
  DEngineInitRespV2,
  DEngineMode,
  DEngineV2,
  DLink,
  DNodeTypeV2,
  DStoreV2,
  DVault,
  EngineDeleteOptsV2,
  EngineUpdateNodesOptsV2,
  EngineWriteOptsV2,
  GetNoteOptsV2,
  GetNotePayloadV2,
  NoteChangeEntry,
  NotePropsDictV2,
  NotePropsV2,
  NoteUtilsV2,
  QueryNotesOpts,
  RenameNoteOptsV2,
  RenameNotePayload,
  RespV2,
  SchemaModuleDictV2,
  SchemaModulePropsV2,
  SchemaQueryResp,
  WorkspaceOpts,
  WriteNoteResp,
} from "@dendronhq/common-all";
import {
  createLogger,
  DLogger,
  readYAML,
  VaultUtils,
  writeYAML,
} from "@dendronhq/common-server";
import _ from "lodash";
import { DConfig } from "./config";
import { FileStorageV2 } from "./drivers/file/storev2";
import { FuseEngine } from "./fuseEngine";

type DendronEngineOptsV2 = {
  wsRoot: string;
  vaultsv3: DVault[];
  forceNew?: boolean;
  store?: any;
  mode?: DEngineMode;
  logger?: DLogger;
};
type DendronEnginePropsV2 = Required<DendronEngineOptsV2>;

export class DendronEngineV2 implements DEngineV2 {
  public wsRoot: string;
  public vaults: string[];
  public store: DStoreV2;
  protected props: DendronEnginePropsV2;
  public logger: DLogger;
  public fuseEngine: FuseEngine;
  public links: DLink[];
  public vaultsv3: DVault[];
  public configRoot: string;

  static _instance: DendronEngineV2 | undefined;

  constructor(props: DendronEnginePropsV2) {
    this.wsRoot = props.wsRoot;
    this.configRoot = props.wsRoot;
    this.store = props.store;
    this.logger = props.logger;
    this.props = props;
    this.fuseEngine = new FuseEngine({});
    this.links = [];
    this.vaultsv3 = props.vaultsv3;
    this.vaults = props.vaultsv3.map((ent) => ent.fsPath);
  }

  static createV3({
    vaults,
    wsRoot,
    logger,
  }: WorkspaceOpts & { logger?: DLogger }) {
    const LOGGER = logger || createLogger();
    return new DendronEngineV2({
      wsRoot,
      vaultsv3: vaults,
      forceNew: true,
      store: new FileStorageV2({
        wsRoot,
        vaultsv3: vaults,
        logger: LOGGER,
      }),
      mode: "fuzzy",
      logger: LOGGER,
    });
  }

  static instance({ vaults, wsRoot }: { vaults: DVault[]; wsRoot: string }) {
    if (!DendronEngineV2._instance) {
      DendronEngineV2._instance = DendronEngineV2.createV3({ vaults, wsRoot });
    }
    return DendronEngineV2._instance;
  }

  get notes(): NotePropsDictV2 {
    return this.store.notes;
  }
  get schemas(): SchemaModuleDictV2 {
    return this.store.schemas;
  }

  /**
   * Does not throw error but returns it
   */
  async init(): Promise<DEngineInitRespV2> {
    try {
      const { data, error } = await this.store.init();
      const { notes, schemas } = data;
      this.updateIndex("note");
      this.updateIndex("schema");
      return {
        error,
        data: { notes, schemas },
      };
    } catch (error) {
      const { message, stack, msg, status, friendly } = error;
      let payload = { message, stack };
      return {
        error: new DendronError({ payload, msg, status, friendly }),
        data: {
          notes: {},
          schemas: {},
        },
      };
    }
  }

  async deleteNote(
    id: string,
    opts?: EngineDeleteOptsV2
  ): ReturnType<DEngineClientV2["deleteNote"]> {
    try {
      const note = this.notes[id];
      const changed = await this.store.deleteNote(id, opts);
      const noteChangeEntry = _.find(
        changed,
        (ent) => ent.note.id === id
      ) as NoteChangeEntry;
      if (noteChangeEntry.status === "delete") {
        await this.fuseEngine.removeNoteFromIndex(note);
      }
      return {
        data: changed,
        error: null,
      };
    } catch (err) {
      return {
        data: [],
        error: err,
      };
    }
  }

  async deleteSchema(
    id: string,
    opts?: EngineDeleteOptsV2
  ): Promise<DEngineDeleteSchemaRespV2> {
    try {
      const data = (await this.store.deleteSchema(
        id,
        opts
      )) as DEngineDeleteSchemaRespV2;
      // deleted schema might affect notes
      await this.updateIndex("note");
      await this.updateIndex("schema");
      return data;
      // FIXM:E not performant
      // const smod = this.schemas[id];
      // await this.fuseEngine.removeSchemaFromIndex(smod);
      // return {
      //   data: undefined,
      //   error: null,
      // };
    } catch (err) {
      return {
        data: {
          notes: {},
          schemas: {},
        },
        error: err,
      };
    }
  }

  async getNoteByPath({
    npath,
    createIfNew,
    vault,
  }: GetNoteOptsV2): Promise<RespV2<GetNotePayloadV2>> {
    const ctx = "getNoteByPath";
    this.logger.debug({ ctx, npath, createIfNew, msg: "enter" });
    const maybeNote = NoteUtilsV2.getNoteByFname(npath, this.notes, {
      vault: vault,
    });
    this.logger.debug({ ctx, maybeNote, msg: "post-query" });
    let noteNew: NotePropsV2 | undefined = maybeNote;
    let changed: NoteChangeEntry[] = [];
    let error = null;
    if ((!maybeNote || maybeNote.stub) && createIfNew) {
      this.logger.debug({ ctx, maybeNote, msg: "create-new" });
      if (maybeNote?.stub) {
        noteNew = maybeNote;
        delete noteNew.stub;
      } else {
        noteNew = NoteUtilsV2.createWithSchema({
          noteOpts: { fname: npath, vault },
          engine: this,
        });
      }
      changed = (await this.writeNote(noteNew, { newNode: true })).data;
    }
    if (!createIfNew && !maybeNote) {
      error = new DendronError({ status: "no_note_found" });
    }
    return {
      data: { note: noteNew, changed },
      error,
    };
  }

  async getConfig() {
    const cpath = DConfig.configPath(this.configRoot);
    const config = readYAML(cpath) as DendronConfig;
    return {
      error: null,
      data: config,
    };
  }

  async getSchema(id: string): Promise<RespV2<SchemaModulePropsV2>> {
    const ctx = "getSchema";
    const data = this.schemas[id];
    this.logger.info({ ctx, msg: "exit" });
    return {
      data,
      error: null,
    };
  }

  queryNotesSync({
    qs,
  }: {
    qs: string;
  }): ReturnType<DEngineClientV2["queryNotesSync"]> {
    const items = this.fuseEngine.queryNote({ qs });
    return {
      error: null,
      data: items.map((ent) => this.notes[ent.id]),
    };
  }

  async querySchema(queryString: string): Promise<SchemaQueryResp> {
    const ctx = "querySchema";

    let items: SchemaModulePropsV2[] = [];
    const results = await this.fuseEngine.querySchema({ qs: queryString });
    items = results.map((ent) => this.schemas[ent.id]);
    // if (queryString === "") {
    //   items = [this.schemas.root];
    // } else if (queryString === "*") {
    //   items = _.values(this.schemas);
    // } else {
    //   const results = this.schemaIndex.search(queryString);
    //   items = _.map(results, (resp) => this.schemas[resp.item.id]);
    // }
    this.logger.info({ ctx, msg: "exit" });
    return {
      error: null,
      data: items,
    };
  }

  async queryNotes(
    opts: QueryNotesOpts
  ): ReturnType<DEngineClientV2["queryNotes"]> {
    const ctx = "Engine:queryNotes";
    const { qs, vault, createIfNew } = opts;
    let items = await this.fuseEngine.queryNote({ qs });
    let item = this.notes[items[0].id];
    if (createIfNew) {
      let noteNew: NotePropsV2;
      if (item?.fname === qs && item?.stub) {
        noteNew = item;
        noteNew.stub = false;
      } else {
        if (_.isUndefined(vault)) {
          return {
            error: new DendronError({ msg: "no vault specified" }),
            data: null as any,
          };
        }
        noteNew = NoteUtilsV2.create({ fname: qs, vault });
      }
      await this.writeNote(noteNew, { newNode: true });
    }
    this.logger.info({ ctx, msg: "exit" });
    let notes = items.map((ent) => this.notes[ent.id]);
    if (!_.isUndefined(vault)) {
      notes = notes.filter((ent) => VaultUtils.isEqual(vault, ent.vault));
    }
    return {
      error: null,
      data: notes,
    };
  }

  // async query(
  //   queryString: string,
  //   mode: DNodeTypeV2,
  //   opts?: QueryOptsV2
  // ): Promise<EngineQueryNoteResp> {
  //   const ctx = "Engine:query";
  //   const cleanOpts = _.defaults(opts || {}, {
  //     fullNode: false,
  //     createIfNew: false,
  //     initialQuery: false,
  //     stub: false,
  //   });
  //   this.logger.info({ ctx, msg: "enter" });
  //   let items: DNodePropsV2[] = [];

  //   // ~~~ schema query
  //   if (mode === "schema") {
  //     throw Error("engine.query for schema is not supported");
  //   } else {
  //     // ~~~ note query
  //     items = await this.fuseEngine.queryNote({ qs: queryString });
  //     // if (queryString === "") {
  //     //   items = [this.notes.root];
  //     // } else if (queryString === "*") {
  //     //   items = _.values(this.notes);
  //     // } else {
  //     //   const results = this.notesIndex.search(queryString);
  //     //   items = _.map(results, (resp) => resp.item);
  //     // }
  //     if (cleanOpts.createIfNew) {
  //       let noteNew: NotePropsV2;
  //       if (items[0]?.fname === queryString && items[0]?.stub) {
  //         noteNew = items[0];
  //         noteNew.stub = false;
  //       } else {
  //         noteNew = NoteUtilsV2.create({ fname: queryString });
  //       }
  //       await this.writeNote(noteNew, { newNode: true });
  //     }
  //     if (cleanOpts.fullNode) {
  //       throw Error("not implemented");
  //     }
  //   }

  //   // ~~~ exit
  //   this.logger.info({ ctx, msg: "exit" });
  //   return {
  //     error: null,
  //     data: items,
  //   };
  // }

  async sync() {
    throw Error("sync not implemented");
    return {} as any;
  }

  async renameNote(opts: RenameNoteOptsV2): Promise<RespV2<RenameNotePayload>> {
    const resp = await this.store.renameNote(opts);
    return {
      error: null,
      data: resp,
    };
  }

  async updateNote(
    note: NotePropsV2,
    opts?: EngineUpdateNodesOptsV2
  ): Promise<void> {
    const out = this.store.updateNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async updateIndex(mode: DNodeTypeV2) {
    if (mode === "schema") {
      this.fuseEngine.updateSchemaIndex(this.schemas);
    } else {
      this.fuseEngine.updateNotesIndex(this.notes);
    }
  }

  async updateSchema(schemaModule: SchemaModulePropsV2) {
    const out = await this.store.updateSchema(schemaModule);
    await this.updateIndex("schema");
    return out;
  }

  async writeConfig(
    opts: ConfigWriteOpts
  ): ReturnType<DEngineV2["writeConfig"]> {
    const { configRoot } = this;
    const cpath = DConfig.configPath(configRoot);
    writeYAML(cpath, opts.config);
    return {
      error: null,
    };
  }

  async writeNote(
    note: NotePropsV2,
    opts?: EngineWriteOptsV2
  ): Promise<WriteNoteResp> {
    const out = await this.store.writeNote(note, opts);
    await this.updateIndex("note");
    return out;
  }

  async writeSchema(schema: SchemaModulePropsV2) {
    return this.store.writeSchema(schema);
  }
}
