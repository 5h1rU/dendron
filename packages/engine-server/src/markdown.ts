// @ts-ignore
import markdownItAST from "markdown-it-ast";
import markdownItRegex from "markdown-it-regex";
import Token from "markdown-it/lib/token";
import _markdownit from "markdown-it";

const markdownIt = _markdownit();

export type MDNode = {
  nodeType: "heading" | "other";
  openNode: Token;
  closeNode: Token;
  children: Token[];
};

export function md2MDNodes(txt: string): MDNode[] {
  const tokens: Token[] = markdownIt
    .use(markdownItRegex, {
      name: "ref-document",
      regex: /\[\[([^\[\]]+?)\]\]/,
      // replace: (rawRef: string) => {
      //   const { ref, label } = parseRef(rawRef);
      //   // const fsPath = findUriByRef(getWorkspaceCache().allUris, ref)?.fsPath;
      //   // if (!fsPath) {
      //   //   return getInvalidRefAnchor(label || ref);
      //   // }
      //   return getRefAnchor(ref, label || ref);
      // },
    })
    .parse(txt, {});
  return markdownItAST.makeAST(tokens);
}
