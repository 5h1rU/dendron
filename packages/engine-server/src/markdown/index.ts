export { RemarkUtils } from "./remark/utils";
export { wikiLinks, WikiLinksOpts } from "./remark/wikiLinks";
export * from "./types";
export * from "./utils";
