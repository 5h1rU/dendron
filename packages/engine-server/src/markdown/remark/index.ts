export * from "./dendronPub";
export * from "./hierarchies";
export * from "./noteRefs";
export * from "./transformLinks";
export { LinkUtils, RemarkUtils } from "./utils";
export { wikiLinks, WikiLinksOpts, matchWikiLink } from "./wikiLinks";
