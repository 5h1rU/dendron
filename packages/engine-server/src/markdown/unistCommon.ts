import { Processor } from "unified";
import { MDUtilsV5 } from ".";
import { DConfig } from "../config";

export class UnistCommonUtils {
  /**
   * Checks whether we should automatically add the FM title as the H1 Title
   * Disabled if inside a note reference
   * @param proc
   */
  static shouldInsertTitle = (proc: Processor) => {
    const { insideNoteRef, config } = MDUtilsV5.getProcData(proc);
    const shouldInsertTitle = insideNoteRef ? false : config.useFMTitle;

    if (!shouldInsertTitle) {
      // NOTE: currently our type definition has this as an optional property
      return DConfig.genDefaultConfig().useFMTitle!;
    } else {
      return shouldInsertTitle;
    }
  };
}
