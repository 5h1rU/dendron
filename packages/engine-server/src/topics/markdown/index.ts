export * from "./utils";
export * from "./utilsv2";
export * from "./plugins";
