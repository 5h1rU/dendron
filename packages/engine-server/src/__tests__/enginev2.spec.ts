import {
  DendronError,
  DEngineV2,
  DVault,
  ENGINE_ERROR_CODES,
  NoteUtilsV2,
  SchemaUtilsV2,
  WorkspaceOpts,
} from "@dendronhq/common-all";
import {
  createLogger,
  note2File,
  readYAML,
  schemaModuleOpts2File,
  SchemaParserV2 as cSchemaParserV2,
} from "@dendronhq/common-server";
import {
  AssertUtils,
  EngineTestUtilsV2,
  ENGINE_PRESETS,
  ENGINE_SERVER,
  INIT_TEST_PRESETS,
  NodeTestPresetsV2,
  NodeTestUtilsV2,
  NoteTestUtilsV3,
  RENAME_TEST_PRESETS,
  runEngineTest,
  runEngineTestV4,
  runJestHarness,
} from "@dendronhq/common-test-utils";
import fs from "fs-extra";
import _ from "lodash";
import path, { toNamespacedPath } from "path";
import { DendronEngineV2 } from "../enginev2";
import { ParserUtilsV2 } from "../topics/markdown";

const { SCHEMAS, NOTES } = ENGINE_SERVER.ENGINE_SINGLE_TEST_PRESET;
const { ENGINE_UPDATE_PRESETS } = ENGINE_SERVER;
const { normalizeNote, normalizeNotes } = NodeTestUtilsV2;

const createEngine = ({ vaults, wsRoot }: WorkspaceOpts) => {
  const logger = createLogger("testLogger", "/tmp/engine-server.txt");
  const engine = DendronEngineV2.createV3({ vaults, wsRoot, logger });
  return engine;
};

const createNotes = async (opts: { rootName: string; vaultDir: string }) => {
  const { rootName, vaultDir } = opts;
  const vault = { fsPath: vaultDir };
  const foo = NoteUtilsV2.create({
    fname: `${rootName}`,
    id: `${rootName}`,
    created: "1",
    updated: "1",
    children: ["ch1"],
    vault,
  });
  const ch1 = NoteUtilsV2.create({
    fname: `${rootName}.ch1`,
    id: `${rootName}.ch1`,
    created: "1",
    updated: "1",
    vault,
  });
  await note2File(foo, vaultDir);
  await note2File(ch1, vaultDir);
  return { foo, ch1 };
};

const beforePreset = async () => {
  const wsRoot = "";
  const vaultDir = await EngineTestUtilsV2.setupVault({
    initDirCb: async (vaultPath: string) => {
      await NodeTestUtilsV2.createSchemas({ vaultPath });
      await NodeTestUtilsV2.createNotes({ vaultPath });
      await NodeTestUtilsV2.createNoteProps({ vaultPath, rootName: "foo" });
      await NodeTestUtilsV2.createSchemaModuleOpts({
        vaultDir: vaultPath,
        rootName: "foo",
      });
    },
  });
  const engine = DendronEngineV2.createV3({
    wsRoot,
    vaults: [{ fsPath: vaultDir }],
  });
  return { vaultDir, engine };
};

describe("engine, schema/", () => {
  let vaultDir: string;
  let engine: DEngineV2;
  let wsRoot = "";

  describe("import", () => {
    let vaultDir: string;
    let engine: DEngineV2;

    beforeEach(async () => {
      vaultDir = await EngineTestUtilsV2.setupVault({
        initDirCb: async (dirPath: string) => {
          const vault = { fsPath: dirPath };
          await NodeTestUtilsV2.createSchemas({ vaultPath: dirPath });
          await NodeTestUtilsV2.createNotes({ vaultPath: dirPath });

          await NodeTestUtilsV2.createSchemaModuleOpts({
            vaultDir: dirPath,
            rootName: "bar",
          });

          let fname = "foo";
          const fooSchema = SchemaUtilsV2.create({
            fname,
            id: fname,
            parent: "root",
            created: "1",
            updated: "1",
            children: ["bar.bar", "baz.baz"],
            vault,
          });
          let module = SchemaUtilsV2.createModule({
            version: 1,
            schemas: [fooSchema],
            imports: ["bar", "baz"],
          });
          await schemaModuleOpts2File(module, dirPath, fname);

          fname = "baz";
          const bazSchema = SchemaUtilsV2.create({
            fname,
            id: fname,
            parent: "root",
            created: "1",
            updated: "1",
            children: ["bar.bar", "ns"],
            vault,
          });
          let childSchema = SchemaUtilsV2.create({
            id: "ns",
            namespace: true,
            vault,
          });
          module = SchemaUtilsV2.createModule({
            version: 1,
            schemas: [bazSchema, childSchema],
            imports: ["bar"],
          });
          await schemaModuleOpts2File(module, dirPath, fname);
        },
      });
      engine = DendronEngineV2.createV3({
        wsRoot,
        vaults: [{ fsPath: vaultDir }],
      });
    });

    test("basic", async () => {
      const { error } = await engine.init();
      expect(error).toBe(null);
      expect(_.size(engine.schemas["foo"].schemas)).toEqual(7);
      expect(_.size(engine.schemas["bar"].schemas)).toEqual(2);
      expect(_.size(engine.schemas["baz"].schemas)).toEqual(4);
      const vault = { fsPath: vaultDir };
      await engine.writeNote(
        NoteUtilsV2.create({ id: "foo.bar", fname: "foo.bar", vault })
      );
      const note = engine.notes["foo.bar"];
      expect(note.schema).toEqual({
        moduleId: "foo",
        schemaId: "bar.bar",
      });
    });

    test("double import", async () => {
      const { error } = await engine.init();
      expect(error).toBe(null);
      const vault = { fsPath: vaultDir };
      await engine.writeNote(
        NoteUtilsV2.create({ id: "foo.baz.bar", fname: "foo.baz.bar", vault })
      );
      const note = engine.notes["foo.baz.bar"];
      expect(note.schema).toEqual({
        moduleId: "foo",
        schemaId: "baz.bar.bar",
      });
    });

    test("import and namespace", async () => {
      const { error } = await engine.init();
      expect(error).toBe(null);
      const vault = { fsPath: vaultDir };
      await engine.writeNote(
        NoteUtilsV2.create({
          id: "foo.baz.ns.one",
          fname: "foo.baz.ns.one",
          vault,
        })
      );
      const note = engine.notes["foo.baz.ns.one"];
      expect(note.schema).toEqual({
        moduleId: "foo",
        schemaId: "baz.ns",
      });
    });
  });
});

describe.only("engine, schemas/", () => {
  const nodeType = "SCHEMAS";

  ENGINE_PRESETS.forEach((pre) => {
    const { name, presets } = pre;
    describe(name, () => {
      test.each(
        _.map(presets[nodeType], (v, k) => {
          return [k, v];
        })
      )("%p", async (_key, TestCase) => {
        const { testFunc, ...opts } = TestCase;
        await runEngineTestV4(testFunc, { ...opts, createEngine, expect });
      });
    });
  });
});

describe("engine, notes/", () => {
  const nodeType = "NOTES";

  ENGINE_PRESETS.forEach((pre) => {
    const { name, presets } = pre;
    describe(name, () => {
      test.each(
        _.map(presets[nodeType], (v, k) => {
          return [k, v];
        })
      )("%p", async (_key, TestCase) => {
        const { testFunc, ...opts } = TestCase;
        await runEngineTestV4(testFunc, { ...opts, createEngine, expect });
      });
    });
  });
});

describe("note and schema", async () => {
  let vaultDir: string;
  let engine: DEngineV2;
  let wsRoot = "";

  beforeEach(async () => {
    vaultDir = await EngineTestUtilsV2.setupVault({
      initDirCb: async (dirPath: string) => {
        await NodeTestUtilsV2.createSchemas({ vaultPath: dirPath });
        await NodeTestUtilsV2.createNotes({ vaultPath: dirPath });
      },
    });
    engine = DendronEngineV2.createV3({
      wsRoot,
      vaults: [{ fsPath: vaultDir }],
    });
  });

  describe("basics/", () => {
    beforeEach(async () => {
      vaultDir = await EngineTestUtilsV2.setupVault({
        initDirCb: async (dirPath: string) => {
          await NodeTestUtilsV2.createSchemas({ vaultPath: dirPath });
          await NodeTestUtilsV2.createNotes({ vaultPath: dirPath });
        },
      });
      engine = DendronEngineV2.createV3({
        wsRoot,
        vaults: [{ fsPath: vaultDir }],
      });
    });

    test("root and two notes", async () => {
      await createNotes({ rootName: "foo", vaultDir });
      await NodeTestUtilsV2.createSchemaModuleOpts({
        vaultDir,
        rootName: "foo",
      });
      await engine.init();
      expect(normalizeNotes(engine.notes)).toMatchSnapshot();
      expect(_.values(engine.notes).length).toEqual(3);
      expect(engine.notes["foo"].schema).toEqual({
        schemaId: "foo",
        moduleId: "foo",
      });
      expect(engine.notes["foo.ch1"].schema).toEqual({
        schemaId: "ch1",
        moduleId: "foo",
      });
    });
  });
});
