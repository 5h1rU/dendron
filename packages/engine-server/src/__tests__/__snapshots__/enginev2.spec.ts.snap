// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engine, notes/ basics bad parse  1`] = `"{\\"opts\\":{\\"status\\":\\"bad_parse_for_note\\",\\"msg\\":\\"{\\\\\\"fname\\\\\\":\\\\\\"foo.md\\\\\\",\\\\\\"error\\\\\\":\\\\\\"can not read a block mapping entry; a multiline key may not be an implicit key at line 5, column 1:\\\\\\\\n    \\\\\\\\n    ^\\\\\\"}\\"},\\"status\\":\\"bad_parse_for_note\\"}"`;

exports[`engine, notes/ basics root 1`] = `
Object {
  "root": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "root",
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
  },
}
`;

exports[`engine, notes/ basics root and two notes 1`] = `
Object {
  "foo": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "parent": "root",
    "title": "Foo",
    "type": "note",
    "updated": "1",
  },
  "foo.ch1": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "parent": "foo",
    "title": "Ch1",
    "type": "note",
    "updated": "1",
  },
  "root": Object {
    "body": "
",
    "children": Array [
      "foo",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "root",
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
  },
}
`;

exports[`engine, notes/ basics stub note 1`] = `
Array [
  "foo.ch1.md",
  "root.md",
  "root.schema.yml",
]
`;

exports[`engine, notes/ getNoteByPath/ get existing note 1`] = `
Object {
  "changed": Array [],
  "note": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "parent": "root",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": "1",
  },
}
`;

exports[`engine, notes/ getNoteByPath/ get new note 1`] = `
Object {
  "changed": Array [
    Object {
      "note": Object {
        "body": "
",
        "children": Array [
          "foo",
          "381a4b9c-955f-45f6-b213-e49901f146b8",
        ],
        "created": "1",
        "custom": Object {},
        "data": Object {},
        "desc": "",
        "fname": "root",
        "id": "root",
        "parent": null,
        "title": "Root",
        "type": "note",
        "updated": "1",
      },
      "status": "update",
    },
    Object {
      "note": Object {
        "body": "",
        "children": Array [],
        "created": 1602363369065,
        "data": Object {},
        "desc": "",
        "fname": "bar",
        "id": "381a4b9c-955f-45f6-b213-e49901f146b8",
        "parent": "root",
        "title": "Bar",
        "type": "note",
        "updated": 1602363369065,
      },
      "status": "create",
    },
  ],
  "note": Object {
    "body": "",
    "children": Array [],
    "created": 1602363369065,
    "data": Object {},
    "desc": "",
    "fname": "bar",
    "id": "381a4b9c-955f-45f6-b213-e49901f146b8",
    "parent": "root",
    "title": "Bar",
    "type": "note",
    "updated": 1602363369065,
  },
}
`;

exports[`engine, notes/ write/ write note, match schema: notes 1`] = `
Object {
  "foo": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "parent": "root",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": "1",
  },
  "foo.ch1": Object {
    "body": "",
    "children": Array [],
    "created": "1",
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "parent": "foo",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": "1",
  },
  "root": Object {
    "body": "
",
    "children": Array [
      "foo",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "root",
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
  },
}
`;

exports[`engine, notes/ write/ write note, no schema 1`] = `
Object {
  "bar": Object {
    "body": "",
    "children": Array [],
    "created": "1",
    "data": Object {},
    "desc": "",
    "fname": "bar",
    "id": "bar",
    "parent": "root",
    "title": "Bar",
    "type": "note",
    "updated": "1",
  },
  "foo": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "parent": "root",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": "1",
  },
  "foo.ch1": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "parent": "foo",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": "1",
  },
  "root": Object {
    "body": "
",
    "children": Array [
      "foo",
      "bar",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "root",
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
  },
}
`;

exports[`engine, schema/ basics/ root 1`] = `
"version: 1
imports: []
schemas:
  - id: root
    children: []
    title: root
    parent: root
"
`;

exports[`engine, schema/ delete/ delete non-root 1`] = `
Object {
  "root": Object {
    "fname": "root",
    "imports": Array [],
    "root": Object {
      "children": Array [],
      "id": "root",
      "parent": "root",
      "title": "root",
    },
    "schemas": Object {
      "root": Object {
        "body": "",
        "children": Array [],
        "created": 1602363368880,
        "data": Object {},
        "desc": "",
        "fname": "__empty",
        "id": "root",
        "parent": "root",
        "title": "root",
        "type": "schema",
        "updated": 1602363368880,
      },
    },
    "version": 1,
  },
}
`;

exports[`engine, schema/ delete/ delete non-root 2`] = `
Array [
  "foo.ch1.md",
  "foo.md",
  "root.md",
  "root.schema.yml",
]
`;

exports[`engine, schema/ write/ update schema, add new prop to module: new schema file 1`] = `
Object {
  "imports": Array [],
  "schemas": Array [
    Object {
      "children": Array [
        "ch1",
        "ch2",
      ],
      "id": "foo",
      "parent": "root",
      "title": "foo",
    },
    Object {
      "children": Array [],
      "id": "ch1",
      "title": "ch1",
    },
    Object {
      "children": Array [],
      "id": "ch2",
      "title": "ch2",
    },
  ],
  "version": 1,
}
`;

exports[`note and schema basics/ root and two notes 1`] = `
Object {
  "foo": Object {
    "body": "
",
    "children": Array [
      "foo.ch1",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo",
    "id": "foo",
    "parent": "root",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "foo",
    },
    "title": "Foo",
    "type": "note",
    "updated": "1",
  },
  "foo.ch1": Object {
    "body": "
",
    "children": Array [],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.ch1",
    "id": "foo.ch1",
    "parent": "foo",
    "schema": Object {
      "moduleId": "foo",
      "schemaId": "ch1",
    },
    "title": "Ch1",
    "type": "note",
    "updated": "1",
  },
  "root": Object {
    "body": "
",
    "children": Array [
      "foo",
    ],
    "created": "1",
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "root",
    "id": "root",
    "parent": null,
    "title": "Root",
    "type": "note",
    "updated": "1",
  },
}
`;
